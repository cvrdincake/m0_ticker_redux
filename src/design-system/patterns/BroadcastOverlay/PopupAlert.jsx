import { useState, useEffect, useRef, useCallback } from 'react';
import { createPortal } from 'react-dom';
import { getGSAP, shouldUseMotion } from '@/lib/motionGuard';
import styles from './PopupAlert.module.css';

/**
 * PopupAlert - Broadcast overlay pattern for non-intrusive notifications
 * Renders via portal at document.body with proper a11y and motion policies
 * 
 * @param {string} title - Alert title (required)
 * @param {string} message - Alert message content (required)
 * @param {React.ReactNode} icon - Optional icon element
 * @param {boolean} active - Controls visibility (default: false)
 * @param {number} duration - Auto-dismiss duration in ms (default: 5000)
 * @param {function} onDismiss - Callback fired on dismiss/timeout
 */
export const PopupAlert = ({
  title,
  message,
  icon,
  active = false,
  duration = 5000,
  onDismiss
}) => {
  const [isVisible, setIsVisible] = useState(false);
  const [isAnimating, setIsAnimating] = useState(false);
  const alertRef = useRef(null);
  const timeoutRef = useRef(null);
  const isPausedRef = useRef(false);
  const remainingTimeRef = useRef(duration);
  const startTimeRef = useRef(null);

  // Clear any active timeout
  const clearTimeout = useCallback(() => {
    if (timeoutRef.current) {
      window.clearTimeout(timeoutRef.current);
      timeoutRef.current = null;
    }
  }, []);

  // Handle dismiss with exit animation
  const handleDismiss = useCallback(() => {
    if (!isVisible) return;
    
    setIsAnimating(false);
    
    if (shouldUseMotion()) {
      const gsap = getGSAP();
      if (gsap && alertRef.current) {
        gsap.to(alertRef.current, {
          opacity: 0,
          scale: 0.9,
          x: 100,
          duration: parseFloat(getComputedStyle(document.documentElement).getPropertyValue('--duration-base')) || 0.24,
          ease: getComputedStyle(document.documentElement).getPropertyValue('--ease-out') || 'power2.out',
          onComplete: () => {
            setIsVisible(false);
            clearTimeout();
            isPausedRef.current = false;
            remainingTimeRef.current = duration;
            onDismiss?.();
          },
        });
        return;
      }
    }
    
    // Fallback for reduced motion
    setTimeout(() => {
      setIsVisible(false);
      clearTimeout();
      isPausedRef.current = false;
      remainingTimeRef.current = duration;
      onDismiss?.();
    }, 100);
  }, [clearTimeout, duration, onDismiss, isVisible]);

  // Start auto-dismiss timeout
  const startTimeout = useCallback(() => {
    if (duration <= 0 || !onDismiss) return;
    
    startTimeRef.current = Date.now();
    timeoutRef.current = setTimeout(() => {
      if (!isPausedRef.current) {
        handleDismiss();
      }
    }, remainingTimeRef.current);
  }, [duration, onDismiss, handleDismiss]);

  // Handle pause on hover (preserve remaining time)
  const handleMouseEnter = useCallback(() => {
    if (!isPausedRef.current && timeoutRef.current) {
      isPausedRef.current = true;
      const elapsed = Date.now() - startTimeRef.current;
      remainingTimeRef.current = Math.max(0, remainingTimeRef.current - elapsed);
      clearTimeout();
    }
  }, [clearTimeout]);

  // Resume timeout on mouse leave
  const handleMouseLeave = useCallback(() => {
    if (isPausedRef.current) {
      isPausedRef.current = false;
      startTimeout();
    }
  }, [startTimeout]);

  // Handle show/hide with proper animation timing
  useEffect(() => {
    if (active) {
      setIsVisible(true);
      setIsAnimating(true);
      remainingTimeRef.current = duration;
      
      // Entry animation with GSAP
      if (shouldUseMotion() && alertRef.current) {
        const gsap = getGSAP();
        if (gsap) {
          gsap.fromTo(alertRef.current,
            { opacity: 0, scale: 0.9, x: 100 },
            { 
              opacity: 1, 
              scale: 1, 
              x: 0,
              duration: parseFloat(getComputedStyle(document.documentElement).getPropertyValue('--duration-base')) || 0.24,
              ease: getComputedStyle(document.documentElement).getPropertyValue('--ease-out') || 'power2.out',
              onComplete: () => {
                startTimeout();
              }
            }
          );
        } else {
          startTimeout();
        }
      } else {
        startTimeout();
      }
    } else {
      handleDismiss();
    }

    return () => clearTimeout();
  }, [active, duration, startTimeout, handleDismiss, clearTimeout]);

  // Handle Esc key to dismiss
  useEffect(() => {
    const handleKeyDown = (event) => {
      if (event.key === 'Escape' && active) {
        handleDismiss();
      }
    };

    if (active) {
      document.addEventListener('keydown', handleKeyDown);
      return () => document.removeEventListener('keydown', handleKeyDown);
    }
  }, [active, handleDismiss]);

  // Don't render if not visible
  if (!isVisible) return null;

  const alertContent = (
    <div
      ref={alertRef}
      className={`
        ${styles.popupAlert}
        ${isAnimating ? styles.popupAlertEntering : styles.popupAlertExiting}
      `}
      role="alert"
      aria-live="assertive"
      aria-atomic="true"
      data-testid="popup-alert"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <div className={styles.popupAlertContent}>
        {icon && (
          <div className={styles.popupAlertIcon} aria-hidden="true">
            {icon}
          </div>
        )}
        <div className={styles.popupAlertText}>
          <div className={styles.popupAlertTitle}>{title}</div>
          <div className={styles.popupAlertMessage}>{message}</div>
        </div>
        <button
          className={styles.popupAlertDismiss}
          onClick={handleDismiss}
          aria-label="Dismiss alert"
          type="button"
        >
          <svg width="16" height="16" viewBox="0 0 16 16" fill="currentColor">
            <path d="M12.854 4.854a.5.5 0 0 0-.708-.708L8 8.293 3.854 4.146a.5.5 0 1 0-.708.708L7.293 9l-4.147 4.146a.5.5 0 0 0 .708.708L8 9.707l4.146 4.147a.5.5 0 0 0 .708-.708L8.707 9l4.147-4.146z"/>
          </svg>
        </button>
      </div>
    </div>
  );

  // Render via portal to document.body
  return createPortal(alertContent, document.body);
};