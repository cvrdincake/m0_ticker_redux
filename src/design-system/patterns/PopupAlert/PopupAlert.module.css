/* PopupAlert.module.css */
.popup-alert {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background: linear-gradient(135deg, 
    rgba(255,255,255,0.15) 0%, 
    rgba(255,255,255,0.08) 100%
  );
  backdrop-filter: blur(32px);
  border: 1px solid rgba(255,255,255,0.2);
  padding: 40px;
  min-width: 400px;
  max-width: 600px;
  text-align: center;
  box-shadow: 0 20px 60px rgba(0,0,0,0.5);
  z-index: 100;
  animation: popup-enter 600ms cubic-bezier(0.68, -0.55, 0.265, 1.55);
}

@keyframes popup-enter {
  from {
    transform: translate(-50%, -50%) scale(0.7);
    opacity: 0;
  }
  to {
    transform: translate(-50%, -50%) scale(1);
    opacity: 1;
  }
}

.popup-icon {
  width: 48px;
  height: 48px;
  margin: 0 auto 20px;
  background: #ffffff;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 24px;
}

.popup-title {
  font-size: 22px;
  font-weight: 200;
  letter-spacing: 0.15em;
  margin-bottom: 16px;
  color: #ffffff;
  text-transform: uppercase;
}

.popup-message {
  font-size: 14px;
  color: #a3a3a3;
  font-weight: 300;
  line-height: 1.6;
}

@media (prefers-reduced-motion: reduce) {
  .popup-alert {
    animation: none;
  }
}