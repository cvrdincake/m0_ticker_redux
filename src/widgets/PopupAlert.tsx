import PopupAlert from '@/widgets/Popup_Alert';

export default PopupAlert;
