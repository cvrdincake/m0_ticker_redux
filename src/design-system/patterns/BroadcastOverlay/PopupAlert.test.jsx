import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { PopupAlert } from './PopupAlert.jsx';

// Mock ReactDOM.createPortal for testing
vi.mock('react-dom', async () => {
  const actual = await vi.importActual('react-dom');
  return {
    ...actual,
    createPortal: (children) => children
  };
});

// Mock motionGuard
vi.mock('@/lib/motionGuard', () => ({
  getGSAP: vi.fn(() => ({
    to: vi.fn((target, options) => {
      // Call onComplete immediately for tests
      setTimeout(() => options.onComplete?.(), 0);
    }),
    fromTo: vi.fn((target, fromOptions, toOptions) => {
      // Call onComplete immediately for tests
      setTimeout(() => toOptions.onComplete?.(), 0);
    }),
  })),
  shouldUseMotion: vi.fn(() => true),
}));

describe('PopupAlert', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  describe('Rendering', () => {
    it('renders when active is true', () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      expect(screen.getByTestId('popup-alert')).toBeInTheDocument();
      expect(screen.getByText('Test Title')).toBeInTheDocument();
      expect(screen.getByText('Test message')).toBeInTheDocument();
    });

    it('does not render when active is false', () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={false}
          onDismiss={vi.fn()}
        />
      );

      expect(screen.queryByTestId('popup-alert')).not.toBeInTheDocument();
    });

    it('renders with icon when provided', () => {
      const TestIcon = () => <svg data-testid="test-icon" />;
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          icon={<TestIcon />}
          active={true}
          onDismiss={vi.fn()}
        />
      );

      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });

    it('has correct accessibility attributes', () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      const alert = screen.getByTestId('popup-alert');
      expect(alert).toHaveAttribute('role', 'alert');
      expect(alert).toHaveAttribute('aria-live', 'assertive');
      expect(alert).toHaveAttribute('aria-atomic', 'true');
    });

    it('has correct CSS classes for entering state', () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      const alert = screen.getByTestId('popup-alert');
      expect(alert.className).toContain('popupAlertEntering');
    });
  });

  describe('Auto-dismiss functionality', () => {
    it('calls onDismiss after duration', async () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          duration={2000}
          onDismiss={mockOnDismiss}
        />
      );

      // Fast-forward time by duration in act()
      await act(async () => {
        vi.advanceTimersByTime(2000);
        await waitFor(() => {
          expect(mockOnDismiss).toHaveBeenCalledTimes(1);
        });
      });
    });

    it('does not auto-dismiss when duration is 0', async () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          duration={0}
          onDismiss={mockOnDismiss}
        />
      );

      await act(async () => {
        vi.advanceTimersByTime(5000);
      });
      expect(mockOnDismiss).not.toHaveBeenCalled();
    });

    it('does not auto-dismiss when onDismiss is not provided', async () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          duration={2000}
        />
      );

      // Should not throw error
      await act(async () => {
        vi.advanceTimersByTime(2000);
      });
    });
  });

  describe('Hover pause functionality', () => {
    it('pauses timer on mouse enter and resumes on mouse leave', async () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          duration={2000}
          onDismiss={mockOnDismiss}
        />
      );

      const alert = screen.getByTestId('popup-alert');

      await act(async () => {
        // Advance halfway through duration
        vi.advanceTimersByTime(1000);

        // Hover - should pause timer
        fireEvent.mouseEnter(alert);
        
        // Advance past original duration - should not dismiss
        vi.advanceTimersByTime(2000);
      });
      
      expect(mockOnDismiss).not.toHaveBeenCalled();

      await act(async () => {
        // Mouse leave - should resume timer with remaining time
        fireEvent.mouseLeave(alert);
        
        // Advance by remaining time (1000ms) + exit animation
        vi.advanceTimersByTime(1000);
        
        await waitFor(() => {
          expect(mockOnDismiss).toHaveBeenCalledTimes(1);
        });
      });
    });
  });

  describe('Keyboard interaction', () => {
    it('dismisses on Escape key', async () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={mockOnDismiss}
        />
      );

      await act(async () => {
        fireEvent.keyDown(document, { key: 'Escape' });
        
        // Wait for exit animation
        await waitFor(() => {
          expect(mockOnDismiss).toHaveBeenCalledTimes(1);
        });
      });
    });

    it('does not dismiss on other keys', () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={mockOnDismiss}
        />
      );

      fireEvent.keyDown(document, { key: 'Enter' });
      fireEvent.keyDown(document, { key: 'Space' });
      fireEvent.keyDown(document, { key: 'Tab' });

      expect(mockOnDismiss).not.toHaveBeenCalled();
    });

    it('only responds to Escape when active', () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={false}
          onDismiss={mockOnDismiss}
        />
      );

      fireEvent.keyDown(document, { key: 'Escape' });
      expect(mockOnDismiss).not.toHaveBeenCalled();
    });
  });

  describe('Dismiss button', () => {
    it('calls onDismiss when dismiss button is clicked', async () => {
      const mockOnDismiss = vi.fn();
      
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={mockOnDismiss}
        />
      );

      const dismissButton = screen.getByLabelText('Dismiss alert');
      
      await act(async () => {
        fireEvent.click(dismissButton);

        await waitFor(() => {
          expect(mockOnDismiss).toHaveBeenCalledTimes(1);
        });
      });
    });

    it('has correct accessibility attributes', () => {
      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      const dismissButton = screen.getByLabelText('Dismiss alert');
      expect(dismissButton).toHaveAttribute('type', 'button');
      expect(dismissButton).toHaveAttribute('aria-label', 'Dismiss alert');
    });
  });

  describe('State transitions', () => {
    it('changes from entering to exiting state when dismissed', async () => {
      const mockOnDismiss = vi.fn();
      
      const { rerender } = render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={mockOnDismiss}
        />
      );

      const alert = screen.getByTestId('popup-alert');
      expect(alert.className).toContain('popupAlertEntering');

      // Trigger dismiss by changing active to false
      await act(async () => {
        rerender(
          <PopupAlert
            title="Test Title"
            message="Test message"
            active={false}
            onDismiss={mockOnDismiss}
          />
        );

        await waitFor(() => {
          expect(alert.className).toContain('popupAlertExiting');
        });
      });
    });

    it('cleans up timers when unmounted', async () => {
      const mockOnDismiss = vi.fn();
      
      const { unmount } = render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          duration={2000}
          onDismiss={mockOnDismiss}
        />
      );

      // Unmount immediately before timer triggers
      unmount();

      // Advance past duration - should not call onDismiss since component is unmounted
      await act(async () => {
        vi.advanceTimersByTime(3000);
      });
      
      expect(mockOnDismiss).not.toHaveBeenCalled();
    });
  });

  describe('Reduced motion', () => {
    it('applies reduced motion styles when preference is set', () => {
      // Mock reduced motion media query
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation(query => ({
          matches: query === '(prefers-reduced-motion: reduce)',
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });

      render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      const alert = screen.getByTestId('popup-alert');
      expect(alert).toBeInTheDocument();
      // Note: Testing actual CSS animation behavior would require a more sophisticated test environment
    });
  });

  describe('Snapshots', () => {
    it('matches snapshot for default state', () => {
      const { container } = render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          active={true}
          onDismiss={vi.fn()}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with icon', () => {
      const TestIcon = () => <svg data-testid="test-icon" />;
      
      const { container } = render(
        <PopupAlert
          title="Test Title"
          message="Test message"
          icon={<TestIcon />}
          active={true}
          onDismiss={vi.fn()}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for long message', () => {
      const { container } = render(
        <PopupAlert
          title="Long Title"
          message="This is a very long message that should wrap across multiple lines to test the layout and styling of the popup alert component when dealing with longer content."
          active={true}
          onDismiss={vi.fn()}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});