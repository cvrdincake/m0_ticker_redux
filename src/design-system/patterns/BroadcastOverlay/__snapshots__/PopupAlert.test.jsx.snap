// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopupAlert > Snapshots > matches snapshot for default state 1`] = `null`;

exports[`PopupAlert > Snapshots > matches snapshot for long message 1`] = `null`;

exports[`PopupAlert > Snapshots > matches snapshot with icon 1`] = `null`;
